### ISSUE TITLE
Training hangs when using DistributedDataParallel in two pod on nodes 

### ISSUE THREAD

**Issue by boxwh1** [pytorch/pytorch#17703]
Training hangs when using DistributedDataParallel in two pod on nodes.
The problem **occasionally** happens  and can not be reproduced each time. 
The pods were started by k8s using flannel network. The physical nodes are connected by 10Ge ethernet.

Training started normally, but hanged in the middle of training process.
![image](https://user-images.githubusercontent.com/41627739/53853400-4860ec80-4000-11e9-9929-678b7547f630.png)
Then there was no output again. The training process did not exit for servel hours.
We found that the GPU utilization  was always 100% on both two nodes.
![image](https://user-images.githubusercontent.com/41627739/53852922-a096ef00-3ffe-11e9-828c-6d38e8d5dd68.png)
After we added NCCL_DEBUG=WARN before the training command, we found the WARN as follows.
On on pod:
_worker0-0d9gnb:15:91 [3] include/socket.h 398 NCCL WARN Call to **write** failed : Connection reset  by peer.
worker0-0d9gnb:15:91 [3] transport.cu:153 NCCL WARN transport.cu:153 ->2 [Proxy thread error]_
On the other pod:
_worker0-0d9gnb:15:91 [3] include/socket.h 398 NCCL WARN Call to **recv** failed : Connection reset  by peer.
worker0-0d9gnb:15:91 [3] transport.cu:153 NCCL WARN transport.cu:153 ->2 [Proxy thread error]_

## Environment
Training code is pull from https://github.com/pytorch/examples/tree/master/imagenet.
Model is Resnet. We used 8 V100s to train (4 on each node).
Batch size was set 8, 32 or 128 per GPU.
The network rate is about 500 MB/s to synchronize the gradients.

 - PyTorch Version (e.g., 1.0): both 1.0 and 0.4.1
 - OS (e.g., Linux): Ubunut 16.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): None
 - Python version: 3.6
 - CUDA/cuDNN version: CUDA 9.0/cuDNN 7.1.2
 - GPU models and configuration: V100 with NVLink
 - Any other relevant information:

**Comment by ldrmcml** [comment]
Tracking this issue.

**Comment by denru01** [comment]
I encountered the same issue when using 1.1.0a0+84b264b. It happens rarely and randomly. I did not test it on the newest version since the newest version runs very slow (at least 2x slower).

**Comment by posix4e** [comment]
This is fully reproducible to me

**Comment by Oewyn** [comment]
I too am able to reproduce this on pytorch 1.1.0 on 2 nodes with worldsize =4, ranks 0,1 on one machine and 2,3 on another connected via direct 10Gbe connection.

-Paul

**Comment by 372046933** [comment]
@boxwh1 Have you got it work? Could you throw some light on your solution?
